File: include/markdown/parsers/block/paragraph.cpp
#include "paragraph.h"
#include "../../utils.h"
#include "../inline/bold.h"
#include "../inline/image_link.h"
#include "../inline/inline_code.h"
#include "../inline/italic.h"
#include "../inline/strikethrough.h"
#include <regex>

namespace paragraph_block {

bool block_parser::identifier(const std::string &block) const {
  const std::regex pattern(R"(^\w+)");

  return std::regex_search(block, pattern);
}

std::string block_parser::parser(const std::string &block) const {
  auto result = block;

  // inline parsers
  std::vector<std::unique_ptr<inline_parser_interface>> inline_parsers;

  // bold inline parser
  inline_parsers.push_back(std::make_unique<bold::inline_parser>());

  // italic inline parser
  inline_parsers.push_back(std::make_unique<italic::inline_parser>());

  // inline code inline parser
  inline_parsers.push_back(std::make_unique<inline_code::inline_parser>());

  // strikethrough inline parser
  inline_parsers.push_back(std::make_unique<strikethrough::inline_parser>());

  // image link inline parser
  inline_parsers.push_back(std::make_unique<image_link::inline_parser>());

  // parse inline
  result = inline_parse(inline_parsers, result);

  return "<p>" + str_trim(result) + "</p>";
}

}

File: Markdown/Parsers/Html/Parsers/Block/Paragraph.cs
namespace Markdown.Parsers.Html.Parsers.Block;

/// <summary>
/// Paragraph block parser. Parses a Markdown paragraph to an HTML paragraph.
/// </summary>
public class Paragraph : IBlockParser
{
    /// <summary>
    /// Default inline parsers for the paragraph block.
    /// </summary>
    /// <returns></returns>
    public static List<IInlineParser> DefaultInlineParsers()
    {
        return [
            new Inline.Bold(),
            new Inline.Italic(),
            new Inline.Link(),
            new Inline.Code()
        ];
    }

    /// <summary>
    /// Parse a Markdown paragraph block to an HTML paragraph.
    /// </summary>
    /// <param name="block"></param>
    /// <returns></returns>
    public string Parse(string block)
    {
        return Parse(block, DefaultInlineParsers());
    }

    /// <summary>
    /// Parse a Markdown paragraph block to an HTML paragraph.
    /// </summary>
    /// <param name="block"></param>
    /// <param name="inlineParsers"></param>
    /// <returns></returns>
    public string Parse(string block, List<IInlineParser> inlineParsers)
    {
        return $"<p>{InlineParser.Parse(block, inlineParsers)}</p>";
    }
}
